TLS classification

nplanel opened the pull request:
### What does this PR do?

Add a mechanism to tags TLS connections passing seen at the packet layer

### Motivation

We would like to tags our TLS connections aim via static tags "packets.protocol:tls"

### Additional Notes

Dependency on agent-payload PR : https://github.com/DataDog/agent-payload/pull/120
and datadog-agent PR https://github.com/DataDog/datadog-agent/pull/9171

### Describe how to test your changes

1. Start system-probe with the following configuration:

```
system_probe_config:
  log_level: debug
  enable_runtime_compiler: true
network_config:
  enabled: true
  enable_http_monitoring: true
  enable_https_monitoring: true
```

2. Do the following requests: `wget https://httpbin.org/anything/foo`
3. Verify you see the tag `"packets.protocol:tls"` in the tags list and his index in the connection:
```
sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections | tee cnx | jq 'conns[].tags , .tags'
```
(please note your unix-socket path may be different)

jq output should be
```
[]
[
  0
]
[]
...
[
  "packets.protocol:tls"
]
```
`
[
  0
]
`
is conns.Tags array index (`0` pointing to `.tags[0] = "packets.protocol:tls"`)
The last json array contains all aggregated tags from all connections ["packets.protocol:tls"]

You can look at the full connections (pid, addresses, ports) via `cat cnx | jq '[ .conns[] | select( .tags | length > 0 ) ]'`

Packets inscpection can be disabled via
```
system_probe_config:

 disable_packets_inspection: true
```
or by setting environment variable `export DD_DISABLE_PACKETS_INSPECTION=true`

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

nplanel committed:
[network/TLS] add TLS support

s/tls_transaction_t/tls_session_t/ ; document and move http_enqueue_with_tags() in tls.h

fixup tls proto

revert 4.4 try to use perf_output on 4.4 (socket)

move TLS to dns socket filter

fixup gnutls by hooking gnutls_deinit(ctx)

tracer_test: https/TLS tests GnuTLS and OpenSSL via wget and curl

GetCnx() every 2sec

debian 10 have curl binary linked with openssl and gnutls but use only openssl during tls query (there no runtime flag available)
this make harder to map lib and tags, one set of tag should match but not both

tracer tests must run under (windows && npm)

increase waiting time delay

HttpKeyFromConn

new packets classifier probe

add classifier telemerty counting tls flow classifed and failed tail_call()

<4.1.0 not supported

remove classifier_enabled ebpf constant

nplanel committed:
[network/TLS] add TLS support

s/tls_transaction_t/tls_session_t/ ; document and move http_enqueue_with_tags() in tls.h

fixup tls proto

revert 4.4 try to use perf_output on 4.4 (socket)

move TLS to dns socket filter

fixup gnutls by hooking gnutls_deinit(ctx)

tracer_test: https/TLS tests GnuTLS and OpenSSL via wget and curl

GetCnx() every 2sec

debian 10 have curl binary linked with openssl and gnutls but use only openssl during tls query (there no runtime flag available)
this make harder to map lib and tags, one set of tag should match but not both

tracer tests must run under (windows && npm)

increase waiting time delay

HttpKeyFromConn

new packets classifier probe

add classifier telemerty counting tls flow classifed and failed tail_call()

<4.1.0 not supported

remove classifier_enabled ebpf constant

brycekahle commented:
@nplanel It looks like this branch includes the changes from #9171. Can you either target that branch, or rebase with just the TLS changes? That will make it easier to review.

nplanel commented:
> @nplanel It looks like this branch includes the changes from #9171. Can you either target that branch, or rebase with just the TLS changes? That will make it easier to review.

@brycekahle sorry PR base wasn't set

p-lambert commented on pkg/network/ebpf/c/http.h:
```diff
@@ -102,6 +103,20 @@ static __always_inline void http_enqueue(http_transaction_t *http, conn_tuple_t
     }
 }
 
+static __always_inline void add_tags_http_tuple_enqueue(skb_info_t *skb_info, u64 tags, u64 started, u64 classified) {
```
can you move this to `tls.c`?
I find both the function name and location a bit misleading. At first I thought it was a general purpose function that would add tags to any `http_transaction_t` and enqueue it, but then I realized that it is TLS specific.

aside from that, could you perhaps leave a comment about what does this `http_transaction_t` represent? It sort of looks like "synthetic" data as you're effectively hard-coding some of its params.
also, I'm a little bit worried that you're enqueuing a transaction that is missing certain fields (like the `request_method` and the `request_fragment`, which could cause this to be inadvertently dropped on the user space side.

nplanel replied:
that was the initial idea to reuse it, but in the current http code we do that in multiple steps (that is fine)
function name come from a long way, you right I renamed it`http_enqueue_with_tags` and move it in `tls.h`

I did start a doc where I added some specifies of the TLS protocol, but I will add a comment in the code as well
https://docs.google.com/document/d/1gC5goIdSQbVe1j-DpYZC6L0wuuy4180hcN0Yr9b0MUA/edit#

nplanel replied:
I updated this branch with 
* fixup that you ask : https://github.com/DataDog/datadog-agent/commit/4c6f7a43d2898f015f1a98c7741f0859bf516d07
* fix gnutls, we must hook gnutls_deinit() https://github.com/DataDog/datadog-agent/commit/8d6056522a0cbaba200ef227389f98e7a8007b6d
* a fix in the TLS proto : send only the HttpStats with time to first APP packet if no tag exist https://github.com/DataDog/datadog-agent/commit/c9ddb6eb11b50ef732a440209117331eb230ac8f
* curl/wget gnutls libssl tracer_test : https://github.com/DataDog/datadog-agent/commit/24185f52576fc2fe1e08787d5a87e2871c96e52c

p-lambert commented on pkg/network/ebpf/c/ip.h:
```diff
@@ -13,6 +13,15 @@
+#define EPHEMERAL_RANGE_BEG 32768
+#define EPHEMERAL_RANGE_END 60999
+
+static __always_inline int is_ephemeral_port(u16 port) {
+    return port >= EPHEMERAL_RANGE_BEG && port <= EPHEMERAL_RANGE_END;
+}
+
```
👍

p-lambert commented on pkg/network/ebpf/c/tls-types.h:
```diff
@@ -0,0 +1,38 @@
+    __u8 __padding;
+    __u64 started;
+    __u64 classified;
+} tls_transaction_t;
```
nit: I think "transaction" may be a bit of a misnomer for TLS. maybe `tls_session_t`?

nplanel replied:
`tls_session_t` looks good

p-lambert reviewed (changes requested):
Please add an integration test :)
(there is an example in `TestHTTPSViaOpenSSLIntegration` of spinning up an HTTPS server that may be helpful as a template, although in that case you can simply issue the requests from Go instead of `wget`)

brycekahle commented on pkg/network/ebpf/c/http.h:
```diff
@@ -13,7 +14,7 @@ static __always_inline void http_prepare_key(u32 cpu, http_batch_key_t *key, htt
 }
 
-static __always_inline void http_notify_batch(struct pt_regs *ctx) {
+static __always_inline void http_notify_batch(void *ctx) {
```
Why change to a `void *` here?

nplanel replied:
was for perf_output, but you can't use it in 4.4 (socket)
I will revert that

p-lambert commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1539,22 +1545,55 @@ func TestHTTPSViaOpenSSLIntegration(t *testing.T) {
+		name     string
+		fetchCmd []string
+	}{
+		{name: "wget", fetchCmd: []string{"wget", "--no-check-certificate", "-O/dev/null"}},
+		{name: "curl", fetchCmd: []string{"curl", "--http1.1", "-k", "-o/dev/null"}},
```
Thank you for doing this 🙇

nplanel replied:
gnutls support need this fix : https://github.com/DataDog/datadog-agent/pull/9981/commits/999f9aa9fb0f873045ecedafab27c56fa9b8e1c0

p-lambert commented on pkg/network/ebpf/c/tls.h:
```diff
@@ -0,0 +1,129 @@
+    }
+    http->tags |= tags;
+    if (!have_tag) {
+        http_enqueue(http, &skb_info->tup);
```
so one bit of context that I'm missing here: are we using an `http_transaction_t` as a hack to send TLS information to userspace or does a HTTP request actually get sent during a TLS handshake? because if this is a hack that raises several questions to me:
* Is TLS classification supposed to work when HTTP monitoring is disabled?
* Should we have a special handling of these "fake" HTTP transactions on userspace?
* What will happen to L7 monitoring and USM monitoring product when they receive these fake HTTP transactions? (cc @brycekahle )
* I've mentioned this in my previous comment, but these are currently invalid as they're missing a path and a method;

p-lambert replied:
Also, instead of flushing things to userspace for protocol classification it may be more interesting to:
1) Piggy back on the `socket/dns_filter` program instead, as this program is much less bloated than HTTP. (we could just give it a more general purpose name;
2) Consider sharing the `maps/conn_stats` map and adding the tags directly to the map entry;

(and by the way, these changes I'm suggesting are mostly about where to hook your code, but they shouldn't require a lot of code changes 😅 )

nplanel replied:
> so one bit of context that I'm missing here: are we using an `http_transaction_t` as a hack to send TLS information to userspace or does a HTTP request actually get sent during a TLS handshake? because if this is a hack that raises several questions to me:

You have good point, I will try to answer them, but I think we should have quick meeting about with @brycekahle 

> * Is TLS classification supposed to work when HTTP monitoring is disabled?

There are no real http request during TLS handshake, this hack/patch will enqueue an httpStats to send the tag :
 * we can drop this mechanism (send the TLS handshake time), we must enable http{,s} monitoring to support it. 

> * Should we have a special handling of these "fake" HTTP transactions on userspace?
> * What will happen to L7 monitoring and USM monitoring product when they receive these fake HTTP transactions? (cc @brycekahle )
> * I've mentioned this in my previous comment, but these are currently invalid as they're missing a path and a method;

Fake or not, I'm sure today how to present it to the backend (and as today it's not required), as it's per protocol metadata and could be different, I talked a little with @heyronhay but it's broader. I looking to add that in the "notes/features on protocols classification" document

nplanel replied:
> Also, instead of flushing things to userspace for protocol classification it may be more interesting to:
> 
> 1. Piggy back on the `socket/dns_filter` program instead, as this program is much less bloated than HTTP. (we could just give it a more general-purpose name);
> 2. Consider sharing the `maps/conn_stats` map and adding the tags directly to the map entry;
> 
> (and by the way, these changes I'm suggesting are mostly about where to hook your code, but they shouldn't require a lot of code changes 😅 )

Totally agree we should share a connection maps with some features like is_classified, drop_packets, ... 
I was plan to work on that and use tail_call.
dns or http will not make a big difference as tls is in tail_call() so have a limited impact on instructions

robertjli reviewed (approved):
Not sure why Processes is tagged here, I don't see any file that we own. Approving to clear this from our queue.

nplanel commented:
> Not sure why Processes is tagged here, I don't see any file that we own. Approving to clear this from our queue.

this was probably because this PR was based on top of https://github.com/DataDog/datadog-agent/pull/9171 , that trigger pkg/process/check/net* as your PR fixed it, thanks !  https://github.com/DataDog/datadog-agent/pull/10263

nplanel commented on pkg/network/dns/monitor_linux.go:
```diff
@@ -56,6 +56,11 @@ func NewReverseDNS(cfg *config.Config) (ReverseDNS, error) {
 		return nil, err
 	}
+
+	if err := p.Start(); err != nil {
```
@brycekahle @p-lambert Why we never start the manager ?

nplanel replied:
I got it, as we got the pkts on userspace .... so I will implement a classifier probe as we can't have both behavior at the same time

p-lambert commented on pkg/network/dns/ebpf.go:
```diff
@@ -32,6 +41,11 @@ func newEBPFProgram(c *config.Config) (*ebpfProgram, error) {
 
 	mgr := &manager.Manager{
+		Maps: []*manager.Map{
+			{Name: string(probes.ConnMap), MapOptions: manager.MapOptions{PinPath: filepath.Join(probes.TracerPinPath, string(probes.ConnMap))}},
```
FYI you don't need to pin maps if you just need to share them within the same process. Here's an example: https://github.com/DataDog/datadog-agent/blob/2f6c9821ab590afdeaafd45bcc1b9dae00f84013/pkg/network/http/ebpf_ssl.go#L137

brycekahle replied:
I actually think we don't want to pin maps, because that makes them live beyond the process.

nplanel replied:
pin should not persist as we manager.Close(cleanAll) and I cleaned up the /sys/bpf temp dir on close
But I agreed, I shared map/fd between the tracer and the classifier

nplanel committed:
fixup ebpf build

nplanel committed:
fixup ebpf build

nplanel committed:
new ebpf manager

nplanel committed:
new ebpf manager

nplanel commented on pkg/network/ebpf/c/prebuilt/classifier.c:
```diff
@@ -0,0 +1,49 @@
+    __u64 val = 0;
+    LOAD_CONSTANT("classifier_enabled", val);
+    return val == ENABLED;
+}
```
yeah I changed the logic in the way, thanks I removed the ebpf constant

nplanel committed:
rebasefixup

nplanel committed:
rebasefixup

juliogreff committed:
[remote tagger] Don't retry stream when shutting down

We noticed the following error in production, happening multiple times
for the same container:

```
error received trying to start stream: tagger stream not started
```

This happens because the remote tagger tries to reconnect when it should
actually just start shutting down. This commit prevents these retries
from happening, by trying to re-establish a stream at the beginning of a
loop, instead of the end, so it'll only fail once, and not retry.

nplanel committed:
fixup conn.Tags and tracer test

nplanel committed:
fixup conn.Tags and tracer test

nplanel committed:
add runtime classifier

nplanel committed:
add runtime classifier

nplanel committed:
f

nplanel committed:
f

nplanel committed:
fcompiletest

nplanel committed:
fcompiletest

nplanel committed:
encodingtest

nplanel committed:
encodingtest

nplanel committed:
runtimeasset

nplanel committed:
runtimeclassifier

nplanel committed:
runtimeclassifier

nplanel committed:
fixupremoveincdefsh

nplanel committed:
fixupremoveincdefsh

nplanel committed:
fixupremoveincdefsh

nplanel committed:
fixupremoveincdefsh

juliogreff committed:
Start tagger before `api.StartServer`

This addresses a race condition affecting the remote tagger. As
`tagger.SetDefaultTagger` happens inside `common.LoadComponents`, before
that happens the agent is running an empty tagger that will never see
any updates. Since there's a possibility that remote taggers will start
between `api.StartServer` and `common.LoadComponents`, there's a race
condition that essentially breaks the remote tagger completely, as we've
seen with missing tags for both the trace and process agent.

To prevent this race condition, this commit moves the initialization of
the tagger *before* starting the agent's HTTP API.

juliogreff committed:
[tagger] Prevent remote tagger wipe on keep-alive

This prevents a keep-alive sent from the core agent (or any empty
message, for that matter) from wiping the remote tagger's contents if it
was previously in an unready state. Although this would be quite rare
and short lived, we suspect this might've been contributing to missing
tags in data being reported from the process agent after the
introduction of the keep-alive messages.

usamasaqib committed:
Merge branch 'nplanel/tls-classif' of github.com:DataDog/datadog-agent into nplanel/tls-classif

juliogreff committed:
[remote tagger] Always honor `<-ctx.Done()`

In some rare instances where the remote tagger is stopped at the same
time as either the health check or telemetry ticker trigger, the
`select` would not see that `t.Ctx` was already cancelled, and would
carry on trying to re-start the stream. This messes with the shut down
of other agent components, as `Tagger.Stop()` seems to block in that
case.

To make sure that `<-t.ctx.Done()` is always evaluated, we now re-start
the loop on the health check and telemetry ticks.

usamasaqib committed:
code restructuring

usamasaqib committed:
code restructuring

usamasaqib committed:
Tagging tls connections correctly after restructuring

usamasaqib committed:
Tagging tls connections correctly after restructuring

usamasaqib committed:
fixed cipher suite parsing

usamasaqib committed:
fixed cipher suite parsing

usamasaqib committed:
remove incorrect comment

usamasaqib committed:
remove incorrect comment

usamasaqib committed:
removed cipher suite parsing from this PR

usamasaqib committed:
removed cipher suite parsing from this PR

usamasaqib committed:
Copy conn_tuple_t to stack in the subprogram for tls parsing.

The function proto for bpf_map_update_elem defines the second argument
to be of type ARG_PTR_TO_MAP_KEY, which is a pointer on stack of the map
key to be used. On most kernel, we can actually provide a pointer to a
map value as well. However on kernel 4.4 the verifier does not allow
this. This commit copies conn_tuple_t, which is used as the key
in the bpf_map_update_elem call, to the stack.

usamasaqib committed:
Copy conn_tuple_t to stack in the subprogram for tls parsing.

The function proto for bpf_map_update_elem defines the second argument
to be of type ARG_PTR_TO_MAP_KEY, which is a pointer on stack of the map
key to be used. On most kernel, we can actually provide a pointer to a
map value as well. However on kernel 4.4 the verifier does not allow
this. This commit copies conn_tuple_t, which is used as the key
in the bpf_map_update_elem call, to the stack.

usamasaqib committed:
merge conflicts resolved

usamasaqib committed:
added runtime generated files

usamasaqib committed:
added runtime generated files

mx-psi committed:
Revert "[remote tagger] Don't retry stream when shutting down" (#12841)

This reverts commit 22395f6339f151872ec1045f1b99e79231846ca6.

L3n41c committed:
Fix a guard ensuring that the arithmetic expression used in allocation cannot overflow (#12833)

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

djmitche committed:
Generalize reflection-based Reporter for stats (#12750)

jcogilvie committed:
admissioncontroller:  Prepend injected env vars (#12821)

* Prepend injected env vars

* Add release note

* Update releasenotes/notes/admissioncontroller-prepend-env-vars-8d88283457c55d9b.yaml

Co-authored-by: Bryce Eadie <bryce.eadie@datadoghq.com>

* Move release note

* Add comment indicating why we prepend

* Refactor config tests not to depend on order of Env

Co-authored-by: Bryce Eadie <bryce.eadie@datadoghq.com>

lebauce committed:
[CWS] Fix signed division causing compilation errors with clang 13 (#12839)

* Fix signed division causing compilation errors with clang 13

* Sync integrity check

Co-authored-by: Paul Cacheux <paul.cacheux@datadoghq.com>

NouemanKHAL committed:
Add comments to exported members for integrations-tools-and-libraries files (#12818)

mbotarro committed:
collect exit code for exit events; create EventType type; refactor tests (#12834)

collect exit code for exit events

hmahmood committed:
Fix test (#12848)

safchain committed:
add agent version to policy metric (#12844)

Commit:
[install_mac_os] Check for existence of user and group selected for systemwide installation (#12825)

Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

safchain committed:
fix discarder with symlink (#12842)

juliogreff committed:
[remote tagger] `continue` loop on errors

This prevents a later call to `t.stream.Recv` from hanging on a stopped
stream, if the stream could not be started (in this case, due to the
tagger itself being stopped).

modernplumbing committed:
Dump discarders (#12505)

* dump discarders cmd frontend

* dump discarders cmd backend

* adding stats maps

* exporting struct fields,fixing print

* add mock, unnecessary go func

* ran "inv -e security-agent.gen-mocks"

* fix typo, add check to not go over map max size

* add printutils

* fix whitespace

* add go:build and copyright headers to new file

* empty to retrigger gitlab pipeline

* Fix usage of errors in security-agent CLI

* Display path when dumping discarders

* Regenerate Go files from protobuf

Co-authored-by: Sylvain Baubeau <lebauce@gmail.com>

safchain committed:
fix pid discader test (#12843)

paulcacheux committed:
CWS: fix nil ptr policy loader, and `model.EventType` key size (#12857)

* Fix policy loader close if not init

* Fix error display

* Fix key size related to `model.EventType`

KSerrania committed:
[omnibus] Patch psutil to fix automount logs in Docker container with disk check (#12604)

Adds an omnibus patch to psutil to fix an issue in the disk check with that started with the 5.7.4 release of psutil.

dependabot[bot] committed:
Bump github.com/prometheus/procfs from 0.7.3 to 0.8.0 (#12861)

* Bump github.com/prometheus/procfs from 0.7.3 to 0.8.0

Bumps [github.com/prometheus/procfs](https://github.com/prometheus/procfs) from 0.7.3 to 0.8.0.
- [Release notes](https://github.com/prometheus/procfs/releases)
- [Commits](https://github.com/prometheus/procfs/compare/v0.7.3...v0.8.0)

---
updated-dependencies:
- dependency-name: github.com/prometheus/procfs
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Auto-generate go.sum and LICENSE-3rdparty.csv changes

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: dependabot[bot] <dependabot[bot]@users.noreply.github.com>

paulcacheux committed:
CWS: fix protobuf size test (#12859)

* add debug helper

* add diff debug for protobuf len

* Improve names

usamasaqib committed:
classifying tls connection even when handshake is missed

usamasaqib committed:
classifying tls connection even when handshake is missed

usamasaqib committed:
tests for tls classification

usamasaqib committed:
tests for tls classification

usamasaqib committed:
fix defer statement

usamasaqib committed:
fix defer statement

usamasaqib committed:
copyright header added

usamasaqib committed:
rebase
